### DISCUSSION TITLE
Web UI not Usable with Older web Browsers

### DISCUSSION THREAD

**Discussion by martinvahi** [community/community#59707] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

A trend seems to be that people publish binaries at 
GitHub releases page, like

https://github.com/TigerVNC/tigervnc/releases

but those pages are broken from the point of view
of older or otherwise retro-oriented browsers like the

http://kmeleonbrowser.org/

which is a very nice browser for WindowsXP.  On one hand,
staying stuck with old technology limits the possibility of taking
advantage of newer technology, but I do not see, that availability of some 
automatic backwards compatibility mode would limit the use of
newer web browser features. The idea:

```
func_at_server_side_generate_some_labeled_static_HTML_that_displays_download_links_even_in_lynx()
b_bool_browser_is_modern=func_at_client_side_run_some_primitive_JavaScript_that_works_with_really_old_browsers_and_that_checks_what_a_browser_supports()
if (b_bool_browser_is_modern){
    func_at_client_side_JavaScript_launch_the_modern_bells_and_whistles_by_overwriting_parts_of_the_server_side_generated_oldfashioned_HTML()
}

```

The reason for using old software like DOS, WindowsXP, may be even older Windowses,
some  modern retro-stuff like FreeDOS, is that hobby-electronics enthusiasts 
create prototypes anyway and the cheapest way to get components that
also happen to be really well documented (unlike modern components) is
to buy out-of-production old chip leftovers, like old microcontrollers, otherwise
old chips, really on the cheap or may be get them even literally for free from
electronic equipment manufacturing companies that clear their warehouses
from components that they are not able to use in production any more due to
the lack of ability to buy more of them. The various kind of  support software for
thos old components, be it proprietary compilers, proprietary IDEs, 
binary donwload/programming utilities, tend to be also OLD, RUN ON OLD 
OPERATING SYSTEMS. So it makes perfect sense to use a modern computer
for doing the code development, then load the code to the old 
hardware and operating system combination or to some virtual appliance 
that runs the old or modern-retro operating system (FreeDOS) and
then run the original old binaries there for uploading/writing 
or compiling the code. 

Another use case is old industrial equipment refurbishing and repairs by using
the adapter design pattern: write some modern UI, let that Raspberry_Pi
interact with the old electronics and otherwise expensive equipment with
expensiv mechanical parts can still be used after some maintenance.

Another scene is the 

https://www.homebrewcpuring.org/ringhome.html

where some people try to connect those old or selfmade computers
to the IPv4 based WWW.

http://www.go4retro.com/products/64nic/

It would be nice, if all of that hacker community, including such type of people

("GeckOS: a Unix-like 6502 operating system | VCFMW 2019")
https://www.youtube.com/watch?v=jtlAOdJmeDI

could take advantage of the GitHub publishing platform, but a requirement
for this is that there is that backwards compatibility. Not to mention
various IoT-like project ideas. For example, I see no reason, why 
GitHub could not compete with Android Marketplace for some 
niche IoT device class. May be host some jar-files or script code based
jar-file analogues, may be some BASIC files that use some really
portable subset of BASIC that is written by a code generator and
not even meant to be read or edited by humans, but still runs
on wide variety of retro-computers, including Commodore64,
if combined with some platform specific normalization library that
does the same kind of platform normalization that JVM promised to do
at early 2000s, except that in stead of just using Commodore64 era
computers, the IoT devices might be using Raspberry_Pico MCU based
devices that run some "UNIX-resembling" operating system

https://github.com/EtchedPixels/FUZIX

That is to say, without the backwards compatibilty, many
interesting use cases, both, hobbyist and commercial, are
EXCLUDED from GitHub use cases.

Thank You for reading my letter.

**Comment by ComputerNerdKev** [comment]
While the releases problem has annoyed me for a while, today (7th of July 2023) I discovered that the problem has become much bigger - Now Git repos are unable to be browsed via the web at all without GitHub's Javascript enabled!

Using the example from the first post, the index on the main repo page is still shown:

https://github.com/TigerVNC/tigervnc/

But if you try to view any files or directories, they aren't shown on browsers without Javascript support (or with it disabled for privacy/security reasons).

https://github.com/TigerVNC/tigervnc/tree/master/media
https://github.com/TigerVNC/tigervnc/blob/master/BUILDING.txt

Also, **this is not just a problem for people using old browsers or computers!**

Many web browsers still developed and available for modern (most often UNIX-based) operating systems do not support Javascript.

I have always preferred Dillo for browsing GitHub. Even though the website has degraded previously in functionality and presentation when viewed in this browser, the speed and efficiency of the browser make it an obvious (to me) preference over browsers based on Chrome/Chromium or Firefox.

There are also some forks of Dillo that are [actually hosted on GitHub](https://github.com/rodarima/dillo).

But this is far wider an issue than just that browser, here is a list of browsers, used on many Linux etc. platforms, that do not have fully-functional Javascript support:

[NetSurf](http://www.netsurf-browser.org/)
[Links](http://links.twibright.com)
[ELinks](https://github.com/rkd77/elinks/) (GitHub)
[Lynx](https://lynx.invisible-island.net/) ([GitHub](https://github.com/ThomasDickey/lynx-snapshots))
[W3m](https://w3m.sourceforge.net/) ([GitHub](https://github.com/tats/w3m)
[Dillo](http://web.archive.org/web/20211231161928/https://www.dillo.org/)

Due to their minimal system requirements and unique features, these browsers are preferred by a range of users over the vastly more complex and inefficient browsers based on the Chromium or Firefox browser engines. Many of these users clearly use GitHub, because some of the developers are even hosting code here.

But it's not just about browsers either. Many users run script-blocker add-ons such as NoScript to avoid the security, performance, and privacy implications of running unverified Javascript code on their computer. These users should be allowed access to GitHub projects, including the [NoScript Git repo](https://github.com/hackademix/noscript) which is hosted on GitHub, without needing to compromise their protection by adding an exception to allow scripts from the GitHub servers.

There is no justification for requiring Javascript in order to browse a directory tree and view text files on the web. Anyone should know that this functionality has been a basic feature of countless websites using plain HTML since the beginning. If additional features are desired which require Javascript to work effectively, then by all means add them as an option like what was suggested in the first post. But cutting off people who are not interested in those features, who simply desire basic online browsing and viewing of source code files without Javascript, is a lazy and pointless approach. I would not recommend anyone use such a website except out of necessity, and especially where there are as many alternative options as there are for Git hosting.

**Reply by martinvahi** [reply]
Thank You, @ComputerNerdKev , for the answer.
My current comment here might be a bit out of topic, but given Your answer, I think that it makes sense to 
point out some general issues with open source software publishing, independent of the publishing portal.
I also have a little bit of good news, which can be summarized as:

`
    tarfile_with_filename_syntax_that_includes_tarfile_secure_hash_id_est_SHA256 + 
    + P2P-filesharing_like_aMule_eMule_etc +
    + "Napster replacement" in terms of signed lists of those tar-file names.
`

I'm totally biased telling about such tar-files, because I have written one
**inefficient, slow, architecturally failed, pile-of-bloat**-Bash-script
that creates and verifies such tar-files:

https://github.com/martinvahi/mmmv_notes/blob/2ca9c6a297a48c54d2935edd2ea5ccac6a651a8c/mmmv_notes/mmmv_utilities/src/various_Bash_scripts/bin_global/silktorrent_packager_t1.bash#L1

For my excuse for still using it: for me it is feature-complete and tested.
It's pretty old script and I think I know, how to make it faster nowadays,
but it works also as it is. As I just wrote, it's feature-complete.

So, now the bad news. I do not want to bash Microsoft, specially given that
they do publish my code without asking money for it, unlike 
https://sourcehut.org/
that might be cheap for Americans, where a software developer can
expect to get at least 10k$/net monthly, but in Estonia a teacher
gets about 1,5k€/month (net, id est that's after paying taxes) and 
software developer salaries are around 2,5k€/month net at best,
more like 2k€/month for real in corporate world and if one wants
to have 50% time off for studying and for working on one's own projects,
basically, for keeping oneself in shape in terms of software development 
skills, then the real income is more like the 1,5k€/month net, what teachers
get. The income can vary for freelancers, but the requirement for freelancers
is that CODE MUST WORK FOR THE CLIENT OR NO PAY! Essentially it means that
a freelancer has to work hard on various components that are not client project
specific and that all takes time that no client pays a dime for.  That is to say,
I find the SourceHut's ~100$/year pricing not that appealing, specially given
that I just want files to be downloadable and do not care about the rest of the
online services that the SourceHut offers. I already pay for hosting huge files at
https://1fichier.com/

On the other hand, **software developers do die**, which means that 
a paid service account of a dead software developer will go offline, because
dead people do not pay their bills. I would like my software to stay online
after I "get hit by a bus" or by some form of Russian munition. _(I'm an Estonian,
living in Estonia, which borders Russia like Ukraine.)_ It seems to me that the 
SourceForge has been an excellent service in that regard, but the issue with the
SourceForge is that their business model is literally in conflict with the needs 
of software developers. Namely, I have created a Ruby script that
uses the GitHub API for generating a Bash script that downloads/clones
all repositories of a given GitHub user account. 

https://github.com/martinvahi/mmmv_notes/blob/2ca9c6a297a48c54d2935edd2ea5ccac6a651a8c/mmmv_notes/mmmv_utilities/src/various_Ruby_scripts/bin_global/mmmv_github_repos_2_clonescript_bash_t1.rb#L1

Demo-video:
("mmmv github repos 2 clonescript bash t1 rb demo", 2019_03_08)
https://www.youtube.com/watch?v=YsFoOMw2cMk

I wanted to create  a similar script for SourceForge, because I want to have
local copies of the code of various open source software projects, including
their dependencies, which often times tend to be probabilistically available
at the same user account, where the open source project itself is published.

```
    ----side-comment---start----
    I have the concept of "software laboratory",
    which for me menas that in a chemistry or biology
    laboratory there are storage areas for chemicals
    that experimenters an choose for conducting their
    experiments, in electronics laboratories there
    are storage areas for electronic components that
    electronics engineers can use for their protptypes
    and in a "software laboratory" there should be a
    LOCAL STORAGE for software components that HAVE
    BEEN TESTED and that can be used for software
    development. And those components should be
    available independent of the availability    
    of any code publishing portals like GitHub,
    SourceForge, ...
    ----side-comment---end-----
```


```
    ----side-comment---start----
    Another sidecomment is that what regads to the
    GitHub, specially the huge acquisition price, billions of
    dollars, about 7.5 billion it was, I guess, that
    Microsoft used for buying GitHub, then actually,
    feature wise, the Sun Microsystems had a very
    similar software publishing project, Project
    Kenai, which the Oracle inherited from Sun
    Microsystems when it acquired all Sun Microsystems
    projects. The dark joke is that while Oracle tried
    to blackmail 9 billiion $ from Google, id est the
    Goolge-vs-Oracle Android related court case, the
    Oracle literally killed that Project Kenai years
    before Microsoft bought GitHub for roughly the
    same money that Oracle was hoping to get from
    Google. I emphasize that the Project Kenai was way
    ahead of GitHub back in the day and would have
    probably out-done GitHub, had Oracle decided not
    to kill the Project Kenai.
    https://archive.ph/OnkS5
    ----side-comment---end-----
```

```
    https://hbr.org/2018/06/why-microsoft-is-willing-to-pay-so-much-for-github
    ------------------citation----start-------------------------
    In other words, Microsoft is not paying $7.5 billion
    for GitHub for its ability to make money (its financial
    value). It’s paying for the access it gets to the legions
    of developers who use GitHub’s code repository products
    on a daily basis (the company’s strategic value) — so
    they can be guided into the Microsoft developer environment,
    where the real money is made.
    ------------------citation----end---------------------------
    archival copy: https://archive.ph/jnCFB
```

Unfortunately, I can not create a similar script for SourceForge, because
that kind of command-line level automation would directly undermine 
the SourceForge business model: force people to watch advertisements.
The related bug report at SourceForge:

https://sourceforge.net/p/forge/feature-requests/311/
archival copy: https://archive.ph/ASgGg

The ZeroNet people have been worrying about the lack of reliably available
code hosting solution for years and try their best to come up with something
that resembles GitHub, but works on ZeroNet. They call it GitCenter

https://github.com/purplesyringa/gitcenter

The fundamental issue with all "alternative networks", be it some IPv4 overlay networks
like Tor, ZeroNet, I2P, etc. or some "physical overlay networks" that use phone lines
like the old Usenet or FidoNet, is that STANDARDS CHANGE, crypto-algorithms change,
the underlaying networks change. That is to say, plain WWW, ZeroNet, FidoNet, etc.
should be seen as **TEMPORARILY FUNCTIONING CHANNELS** and real software dependency 
graph should never use any network specific or even service specific address strings as 
package ID-s. For example, a project name is unique only in the context/namespace of
a single code-hosting service (GitHub/SourceForge/NPM/etc.) But a tar-file that is 
named according to its secure hash is identified regardless of the channel/service that
is used for transporting it. The transport channel for such tar-files might be even
a drone, a pigeon with a memory card, Soviet Union and U.S.A. Cold War era
"spy-sneakernet", etc. Hence my efforts on working on data collection management
systems in stead of things like ZeroNet, although I started the study of that matter
by trying to study about the ZeroNet and alike.  I collected the various scientific
papers, a considerable amount of source code, some videoclips and other
related materials to a Fossil repository that is about 70GiB and as of 2023_07_07 MIGHT
be downloadable from

(Anty-bot password, "FooBar9897", has to be entered manually, not copy-pasted.)
https://1fichier.com/dir/oEJdfG58

Fossil is like git and Wiki 2 in one. The wiki is served by a local, integrated,
web server.

    https://fossil-scm.org/

The original author of the Fossil is the same as that of SQLite: Richard Hipp.
I strongly recommend to use NilFS2 for holding relevant data, because
the ext4 is not reliable, I lost my main copy of that huge, ~70GiB, file
once on ext4. The NilFS2 does not have error correction for data, only for
filesystem datastructures, but neither does ext4. NilFS2 deletes by
adding "file deleted" notes to log and the disk/SSD space is actually
freed by a daemon that reads the log. That's why I use ext4 partitions
for build folders and NilFS2 partitions for storage. And I use RAM-partitions
for temporary files, because about 60MiB sized RAM partition tends to be
plenty, if a temporary file that only lives a few seconds, is just ~20KiB max.

When it comes to personal data collection, like, clones of repositories,
then I recommend to store them on NilFS2 partitons that reside on
USB HDDs, because motherboard standards change, but an USB storage
device can be connected to various kinds of computers, independent of 
what interfaces the motherboard otherwise supports. Again, the goal is
long-term storage, where magnetic disks are way better than any
Flash memory based devices. For build folders and such, any fast 
bells and whistles that one can get its hands on, is recommended :-)

So, basically, the unique role of a code-hosting site like SourceForge or GitHub is 
to be a channel for data exchange, where the data is code, and to be a
kind of software archive in situations, where the original developers 
are dead, kidnapped, imprisoned in some awful dictatorship, etc.
Those roles do not match with the business goals of Microsoft or
SourceForge holding company. Microsoft owns LinkedIn and GitHub
is just a data milking cow about global software developer "human resources",
except that the true corporate souls, who love to work at big corporations,
often times do not even have GitHub accounts, because they develop closed source
software for megacorporations and therefore DO NOT HAVE any 
significant software that they self own copyrights to. It's usually the
freelancers or otherwise people, who work at small companies,
who publish anything on GitHub, unless the megacorporation
has some "open source policy" or open source projects,
which is NOT the case for most big companies even in 2023.
Speciallty those companies that earn their income from selling
closed source software licenses. (Microsoft as license sales based
megacorporation does not count, because
it has an "open source policy/strategy".) That is to say, the main
sweet-spot for LinkedIn data input would be a list of those software developers, who
actually love to work at big companies, but the LinkedIn/Microsoft is not getting that
data from GitHub or SourceForge or npmjs, because that demographic is not
that active at open source publishing sites. A stereotype of the sweet-spot-demographic for LinkedIn: 
someone with a doctoral degree and no interests in running their own business, because 
there are always many opportunities for interesting technical jobs with 
relatively good pay for such typical "corporate doctors". For comparison,
I'm a looser-lowly-educated-freelancer, who at age 41.9~42 has
only a Bachelors in IT, no interest in working halve one's life in 
one megacorporation and even, when some HR people do want me
to apply, there's no point for me to waste my time applying, because the
offered salary would be a guaranteed poverty in a situation, where I also
lack the freedom to study and lack the freedom to work on my own projects.
Basically, I would really need to eat a lot of antidepressants, if I get a corporate job.

In that sense, may be Microsoft should have not acquired GitHub (they do not
get the data from GitHub that the LinkedIn business is really after) and since 
freelancers like me is not even the target demographic for LinkedIn, there is no
business sense to make GitHub more practical for people like me. But, on a positive note,
may be Microsoft does the unrational move and still tries to make GitHub 
more practical for people like me.... May be I get lucky :-D

Thank You for Your answer and thank You for reading my comment.

P.S. One of my idols is the aerospace engineer Burt Rutan, who also has just a Bachelor's degree, but 
who has clearly achieved way more in life in aerospace engineering than 
the hordes of people, who have a doctoral degree in aerospace engineering.
He also worked as a freelancer. Probably made his best work as a freelancer :-)

**Comment by rv0** [comment]
a +1 here

I have noticed the issue before that the links to binaries are not shown on older browsers.
Today I just re-installed an old Mac Mini with El Capitan (10.11) and wanted to download another browser (Arctic Fox). With the Safari version on it I can't see the download links, basically making one of the most simple features of github ("download a binary") broken.

I understand we can't get full Github functionality on these older browsers, but just a link to a download is something so basic, I find it a big lack of respect of web standards and accessibility that this is broken and remains broken.

A lot of projects release binaries for older systems via GitHub, there really should be an accessible github lite version, or just the default version should gracefully fallback when some of the (javascript?) functionality is not available.

**Reply by martinvahi** [reply]
>... 
> I have noticed the issue before that the links to binaries are not shown on older browsers.
>... 
> A lot of projects release binaries for older systems via GitHub, there really should be an accessible github lite version, or just the default version should gracefully fallback when some of the (javascript?) functionality is not available.

@rv0  You may want to try a Google Chrome fork that has been specially designed to bring modern web features to WindowsXP, the [Supermium](https://github.com/win32ss/supermium/releases). I use it also on Windows10 and Windows11, because that way I am not being dragged to wherever the Google wants to drag me while all of the nice Google Chrome plugins like Vimium, Ublock Origin, etc. keep on working not as an accident or a byproduct, but because the authors of Supermium intentionally keep them supported (Manifest v2 topic). Also there is the GitHub [command line utility](https://cli.github.com/), which is kind of awful to install from source, at least that was the case some time ago, when I tried to install it from source, but it MIGHT be available at a Debian or Debian-like Linux distribution standard package collection. The name of the command line program is "gh".

    which gh # at a Bash console

tells, whether You have it installed already.

Thank You for reading my comment.
